lib/fluent/command/cat.rb

```diff
@@ -20,320 +20,347 @@
-  format = 'msgpack'
-}
-
-op.on('--none', "same as: -f none", TrueClass) {|b|
-  format = 'none'
-}
-
-op.on('--message-key KEY', "key field for none format (default: #{message_key})") {|s|
-  message_key = s
-}
-
-op.on('--time-as-integer', "Send time as integer for v0.12 or earlier", TrueClass) { |b|
-  time_as_integer = true
-}
-
-op.on('--retry-limit N', "Specify the number of retry limit (default: #{retry_limit})", Integer) {|n|
-  retry_limit = n
-}
-
-singleton_class.module_eval do
-  define_method(:usage) do |msg|
-    puts op.to_s
-    puts "error: #{msg}" if msg
-    exit 1
-  end
-end
+module Fluent
+  class Cat
+    def initialize(argv = ARGV)
+      @argv = argv
+      @opt_parser = OptionParser.new
+      configure_option_parser
+    end
 
-begin
-  op.parse!(ARGV)
+    class Writer
+      include MonitorMixin
 
-  if ARGV.length != 1
-    usage nil
-  end
+      RetryLimitError = Class.new(StandardError)
 
-  tag = ARGV.shift
+      class TimerThread
+        def initialize(writer)
+          @writer = writer
+        end
 
-rescue
-  usage $!.to_s
-end
+        def start
+          @finish = false
+          @thread = Thread.new(&method(:run))
+        end
 
-require 'socket'
-require 'yajl'
-require 'msgpack'
-require 'fluent/ext_monitor_require'
+        def shutdown
+          @finish = true
+          @thread.join
+        end
 
-class Writer
-  include MonitorMixin
+        def run
+          until @finish
+            sleep 1
+            @writer.on_timer
+          end
+        end
+      end
 
-  RetryLimitError = Class.new(StandardError)
+      def initialize(tag, connector, time_as_integer: false, retry_limit: 5)
+        @tag = tag
+        @connector = connector
+        @socket = false
 
-  class TimerThread
-    def initialize(writer)
-      @writer = writer
-    end
+        @socket_time = Time.now.to_i
+        @socket_ttl = 10  # TODO
+        @error_history = []
 
-    def start
-      @finish = false
-      @thread = Thread.new(&method(:run))
-    end
+        @pending = []
+        @pending_limit = 1024  # TODO
+        @retry_wait = 1
+        @retry_limit = retry_limit
+        @time_as_integer = time_as_integer
 
-    def shutdown
-      @finish = true
-      @thread.join
-    end
+        super()
+      end
 
-    def run
-      until @finish
-        sleep 1
-        @writer.on_timer
+      def secondary_record?(record)
+        record.class != Hash and
+          record.size == 2 and
+          record.first.class == Fluent::EventTime and
+          record.last.class == Hash
+      end
+
+      def write(record)
+        unless secondary_record?(record)
+          if record.class != Hash
+            raise ArgumentError, "Input must be a map (got #{record.class})"
+          end
+        end
+
+        time = Fluent::EventTime.now
+        time = time.to_i if @time_as_integer
+        entry = if secondary_record?(record)
+                  [time, record.last]
```

Comment:
Shouldn't we reuse EventTime in the original data here?

Reply:
Even though secondary file, it may be better to follow the actual sent time (not original) for consistency.

How do you think @ashie , @cosmo0920 ?

Reply:
I think that throwing original time data is not good. 
How about include original time data under `original_time` or `stored_time` key or something...?

Reply:
There is one concern about posting original timestamp data as an alias `original_time` or `stored_time`, does it conflict with the secondary's record? are there any good key name for it?
To avoid such a situation, not posting it and document timestamp key behavior may be reasonable.
:thinking:

Reply:
Please add a comment for this discussion.